File: examples/node/neutral_ipc_template/NeutralIpcTemplate.js
/**
 * Neutral Node.js IPC client for Neutral TS.
 * https://github.com/FranBarInstance/neutral-ipc
 */

const net = require('net');
const { HOST, PORT, TIMEOUT, BUFFER_SIZE } = require('./NeutralIpcConfig');

class NeutralIpcRecord {

    // ============================================
    // Neutral IPC record version 0 (draft version)
    // ============================================
    //
    // HEADER:
    //
    // \x00              # reserved
    // \x00              # control (action/status) (10 = parse template)
    // \x00              # content-format 1 (10 = JSON, 20 = file path, 30 = plaintext, 40 = binary)
    // \x00\x00\x00\x00  # content-length 1 big endian byte order
    // \x00              # content-format 2 (10 = JSON, 20 = file path, 30 = plaintext, 40 = binary)
    // \x00\x00\x00\x00  # content-length 2 big endian byte order (can be zero)
    //
    // All text utf8


    static RESERVED = 0;
    static HEADER_LEN = 12;
    static CTRL_PARSE_TEMPLATE = 10;
    static CTRL_STATUS_OK = 0;
    static CTRL_STATUS_KO = 1;
    static CONTENT_JSON = 10;
    static CONTENT_PATH = 20;
    static CONTENT_TEXT = 30;
    static CONTENT_BIN = 40;

    static decodeHeader(recordHeader) {
        const reserved = recordHeader[0];
        const control = recordHeader[1];
        const format1 = recordHeader[2];
        const length1 = recordHeader.readUInt32BE(3);
        const format2 = recordHeader[7];
        const length2 = recordHeader.readUInt32BE(8);
        return {
            'reserved': reserved,
            'control': control,
            'format-1': format1,
            'length-1': length1,
            'format-2': format2,
            'length-2': length2,
        };
    }

    static encodeHeader(control, format1, length1, format2, length2) {
        const header = Buffer.alloc(this.HEADER_LEN);
        header[0] = this.RESERVED;
        header[1] = control;
        header[2] = format1;
        header.writeUInt32BE(length1, 3);
        header[7] = format2;
        header.writeUInt32BE(length2, 8);
        return header;
    }

    static encodeRecord(control, format1, content1, format2, content2) {
        const length1 = Buffer.byteLength(content1, 'utf8');
        const length2 = Buffer.byteLength(content2, 'utf8');
        const header = this.encodeHeader(control, format1, length1, format2, length2);
        const content1Buf = Buffer.from(content1, 'utf8');
        const content2Buf = Buffer.from(content2, 'utf8');
        return Buffer.concat([header, content1Buf, content2Buf]);
    }

    static decodeRecord(header, content1, content2) {
        const decodedHeader = this.decodeHeader(header);
        return {
            "reserved": this.RESERVED,
            "control": decodedHeader['control'],
            'format-1': decodedHeader['format-1'],
            'content-1': content1,
            'format-2': decodedHeader['format-2'],
            'content-2': content2,
        };
    }
}

class NeutralIpcClient {
    constructor(control, format1, content1, format2, content2) {
        this.control = control;
        this.format1 = format1;
        this.content1 = content1;
        this.format2 = format2;
        this.content2 = content2;
        this.result = {};
    }

    async start() {
        return new Promise((resolve, reject) => {
            const socket = net.createConnection({ host: HOST, port: PORT, timeout: TIMEOUT }, () => {
                const request = NeutralIpcRecord.encodeRecord(
                    this.control, this.format1, this.content1, this.format2, this.content2
                );
                socket.write(request);
            });

            socket.setTimeout(TIMEOUT);
            socket.on('timeout', () => {
                socket.destroy();
                reject(new Error('Socket timeout'));
            });
            socket.on('error', reject);

            let responseHeader;
            let totalReceived = 0;
            let content1 = '';
            let content2 = '';
            let readingHeader = true;
            let headerReceived = false;
            let length1 = 0;
            let length2 = 0;
            let content1Received = 0;
            let content2Received = 0;

            socket.on('data', (data) => {
                if (readingHeader) {
                    if (!responseHeader) {
                        responseHeader = Buffer.alloc(NeutralIpcRecord.HEADER_LEN);
                    }
                    const toCopy = Math.min(data.length, NeutralIpcRecord.HEADER_LEN - totalReceived);
                    data.copy(responseHeader, totalReceived, 0, toCopy);
                    totalReceived += toCopy;
                    data = data.slice(toCopy);

                    if (totalReceived === NeutralIpcRecord.HEADER_LEN) {
                        const response = NeutralIpcRecord.decodeHeader(responseHeader);
                        length1 = response['length-1'];
                        length2 = response['length-2'];
                        readingHeader = false;
                        headerReceived = true;
                    }
                }

                if (headerReceived) {
                    if (content1Received < length1) {
                        const toCopy = Math.min(data.length, length1 - content1Received);
                        const buf = Buffer.from(data.slice(0, toCopy));
                        content1 += buf.toString('utf8');
                        content1Received += toCopy;
                        data = data.slice(toCopy);
                    }

                    if (content1Received === length1 && content2Received < length2) {
                        const toCopy = Math.min(data.length, length2 - content2Received);
                        const buf = Buffer.from(data.slice(0, toCopy));
                        content2 += buf.toString('utf8');
                        content2Received += toCopy;
                        data = data.slice(toCopy);
                    }

                    if (content1Received === length1 && content2Received === length2) {
                        socket.end();
                        this.result = NeutralIpcRecord.decodeRecord(responseHeader, content1, content2);
                        resolve(this.result);
                    }
                }
            });

            socket.on('end', () => {
                if (!headerReceived || content1Received < length1 || content2Received < length2) {
                    reject(new Error('Incomplete response'));
                }
            });
        });
    }
}

class NeutralIpcTemplate {
    constructor(template, schema, tplType = NeutralIpcRecord.CONTENT_PATH) {
        this.template = template;
        this.tplType = tplType;
        this.schema = typeof schema === 'object' ? JSON.stringify(schema) : schema;
        this.result = {};
    }

    async render() {
        const record = new NeutralIpcClient(
            NeutralIpcRecord.CTRL_PARSE_TEMPLATE,
            NeutralIpcRecord.CONTENT_JSON,
            this.schema,
            this.tplType,
            this.template
        );
        const result = await record.start();
        this.result = {
            'status': result['control'],
            'result': JSON.parse(result['content-1']),
            'content': result['content-2'],
        };
        return this.result['content'];
    }

    setPath(path) {
        this.tplType = NeutralIpcRecord.CONTENT_PATH;
        this.template = path;
    }

    setSource(source) {
        this.tplType = NeutralIpcRecord.CONTENT_TEXT;
        this.template = source;
    }

    mergeSchema(schema) {
        const currentSchema = JSON.parse(this.schema);
        const newSchema = typeof schema === 'string' ? JSON.parse(schema) : schema;
        this.schema = JSON.stringify(deepMerge(currentSchema, newSchema));
    }

    hasError() {
        return this.result.status !== 0 || (this.result.result && this.result.result.has_error);
    }

    getStatusCode() {
        return this.result.result ? this.result.result.status_code : null;
    }

    getStatusText() {
        return this.result.result ? this.result.result.status_text : null;
    }

    getStatusParam() {
        return this.result.result ? this.result.result.status_param : null;
    }

    getResult() {
        return this.result.result || null;
    }
}

function deepMerge(dict1, dict2) {
    const merged = { ...dict1 };
    for (const [key, value] of Object.entries(dict2)) {
        if (merged[key] && typeof merged[key] === 'object' && typeof value === 'object' && !Array.isArray(merged[key]) && !Array.isArray(value)) {
            merged[key] = deepMerge(merged[key], value);
        } else {
            merged[key] = value;
        }
    }
    return merged;
}

module.exports = NeutralIpcTemplate;

File: examples/go/neutral_ipc_template/neutral_ipc_template.go
package neutral_ipc_template

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"net"
	"time"

	msgpack "github.com/vmihailenco/msgpack/v5"
)

// NeutralIpcRecord constants and methods.
const (
	Reserved          = 0
	HeaderLen         = 12
	CtrlParseTemplate = 10
	CtrlStatusOk      = 0
	CtrlStatusKo      = 1
	ContentJSON       = 10
	ContentMsgpack    = 50
	ContentPath       = 20
	ContentText       = 30
	ContentBin        = 40
	Empty             = ""
)

type NeutralIpcRecord struct {
	Reserved byte
	Control  byte
	Format1  byte
	Length1  uint32
	Format2  byte
	Length2  uint32
}

func (r *NeutralIpcRecord) DecodeHeader(recordHeader []byte) error {
	if len(recordHeader) != HeaderLen {
		return fmt.Errorf("invalid header length")
	}
	r.Reserved = recordHeader[0]
	r.Control = recordHeader[1]
	r.Format1 = recordHeader[2]
	r.Length1 = binary.BigEndian.Uint32(recordHeader[3:7])
	r.Format2 = recordHeader[7]
	r.Length2 = binary.BigEndian.Uint32(recordHeader[8:12])
	return nil
}

func (r *NeutralIpcRecord) EncodeHeader() []byte {
	header := make([]byte, HeaderLen)
	header[0] = Reserved
	header[1] = r.Control
	header[2] = r.Format1
	binary.BigEndian.PutUint32(header[3:7], r.Length1)
	header[7] = r.Format2
	binary.BigEndian.PutUint32(header[8:12], r.Length2)
	return header
}

func (r *NeutralIpcRecord) EncodeRecord(control byte, format1 byte, content1 interface{}, format2 byte, content2 string) []byte {
	r.Control = control
	r.Format1 = format1

	var content1Bytes []byte
	switch c := content1.(type) {
	case string:
		content1Bytes = []byte(c)
	case []byte:
		content1Bytes = c
	default:
		content1Bytes = []byte(c.(string))
	}

	r.Length1 = uint32(len(content1Bytes))
	r.Format2 = format2
	r.Length2 = uint32(len(content2))
	header := r.EncodeHeader()
	return append(header, append(content1Bytes, []byte(content2)...)...)
}

func (r *NeutralIpcRecord) DecodeRecord(header []byte, content1, content2 string) map[string]interface{} {
	if err := r.DecodeHeader(header); err != nil {
		return nil
	}
	return map[string]interface{}{
		"reserved":  r.Reserved,
		"control":   r.Control,
		"format-1":  r.Format1,
		"content-1": content1,
		"format-2":  r.Format2,
		"content-2": content2,
	}
}

// NeutralIpcClient handles IPC communication.
type NeutralIpcClient struct {
	control  byte
	format1  byte
	content1 string
	format2  byte
	content2 string
	result   map[string]interface{}
}

func NewNeutralIpcClient(control byte, format1 byte, content1 string, format2 byte, content2 string) *NeutralIpcClient {
	return &NeutralIpcClient{
		control:  control,
		format1:  format1,
		content1: content1,
		format2:  format2,
		content2: content2,
	}
}

func (c *NeutralIpcClient) Start() (map[string]interface{}, error) {
	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", GetHost(), GetPort()), time.Duration(GetTimeout())*time.Second)
	if err != nil {
		return nil, err
	}
	defer conn.Close()

	record := &NeutralIpcRecord{}
	request := record.EncodeRecord(c.control, c.format1, c.content1, c.format2, c.content2)
	if _, err := conn.Write(request); err != nil {
		return nil, err
	}

	responseHeader := make([]byte, HeaderLen)
	if _, err := conn.Read(responseHeader); err != nil {
		return nil, err
	}

	respRecord := &NeutralIpcRecord{}
	if err := respRecord.DecodeHeader(responseHeader); err != nil {
		return nil, err
	}

	content1, err := c.readContent(conn, int(respRecord.Length1))
	if err != nil {
		return nil, err
	}
	content2, err := c.readContent(conn, int(respRecord.Length2))
	if err != nil {
		return nil, err
	}

	c.result = respRecord.DecodeRecord(responseHeader, content1, content2)
	return c.result, nil
}

func (c *NeutralIpcClient) readContent(conn net.Conn, length int) (string, error) {
	if length == 0 {
		return "", nil
	}
	chunks := make([][]byte, 0)
	remaining := length

	for remaining > 0 {
		n := min(GetBufferSize(), remaining)
		chunk := make([]byte, n)
		read := 0
		for read < n {
			chunkRead, err := conn.Read(chunk[read:])
			if err != nil {
				return "", err
			}
			read += chunkRead
		}
		chunks = append(chunks, chunk)
		remaining -= n
	}

	var buffer bytes.Buffer
	for _, chunk := range chunks {
		buffer.Write(chunk)
	}
	return buffer.String(), nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// NeutralIpcTemplate is the main IPC template class.
type NeutralIpcTemplate struct {
	template   string
	tplType    byte
	schemaType byte
	schema     interface{}
	result     map[string]interface{}
}

func NewNeutralIpcTemplate(template string, schema interface{}) *NeutralIpcTemplate {
	return NewNeutralIpcTemplateWithSchemaType(template, schema, ContentJSON)
}

func NewNeutralIpcTemplateWithSchemaType(template string, schema interface{}, schemaType byte) *NeutralIpcTemplate {
	var schemaData interface{}

	if schemaType == ContentMsgpack {
		switch s := schema.(type) {
		case string:
			var parsed interface{}
			if err := json.Unmarshal([]byte(s), &parsed); err == nil {
				data, _ := msgpack.Marshal(parsed)
				schemaData = data
			} else {
				schemaData = []byte{}
			}
		case map[string]interface{}:
			data, _ := msgpack.Marshal(s)
			schemaData = data
		default:
			schemaData = []byte{}
		}
	} else {
		if s, ok := schema.(string); ok {
			schemaData = s
		} else if sch, ok := schema.(map[string]interface{}); ok {
			data, _ := json.Marshal(sch)
			schemaData = string(data)
		}
	}

	return &NeutralIpcTemplate{
		template:   template,
		tplType:    ContentPath,
		schemaType: schemaType,
		schema:     schemaData,
	}
}

func (t *NeutralIpcTemplate) Render() string {
	var schemaStr string
	switch s := t.schema.(type) {
	case string:
		schemaStr = s
	case []byte:
		schemaStr = string(s)
	default:
		schemaStr = ""
	}

	client := NewNeutralIpcClient(CtrlParseTemplate, t.schemaType, schemaStr, t.tplType, t.template)
	result, err := client.Start()
	if err != nil {
		t.result = map[string]interface{}{"error": err.Error()}
		return ""
	}
	t.result = result

	status, ok := result["control"].(byte)
	if !ok {
		t.result["error"] = "invalid control type in result"
		return ""
	}
	content1, ok := result["content-1"].(string)
	if !ok {
		t.result["error"] = "invalid content-1 type in result"
		return ""
	}
	content2, ok := result["content-2"].(string)
	if !ok {
		t.result["error"] = "invalid content-2 type in result"
		return ""
	}

	var res map[string]interface{}
	if err := json.Unmarshal([]byte(content1), &res); err != nil {
		t.result["result"] = content1
	} else {
		t.result["result"] = res
	}
	t.result["status"] = status
	t.result["content"] = content2

	return t.result["content"].(string)
}

func (t *NeutralIpcTemplate) SetPath(path string) {
	t.tplType = ContentPath
	t.template = path
}

func (t *NeutralIpcTemplate) SetSource(source string) {
	t.tplType = ContentText
	t.template = source
}

func (t *NeutralIpcTemplate) MergeSchema(schema interface{}) {
	if t.schemaType == ContentMsgpack {
		return // Cannot merge msgpack schemas
	}

	schemaStr := ""
	if s, ok := schema.(string); ok {
		schemaStr = s
	} else if sch, ok := schema.(map[string]interface{}); ok {
		data, _ := json.Marshal(sch)
		schemaStr = string(data)
	}

	var current map[string]interface{}
	switch s := t.schema.(type) {
	case string:
		if err := json.Unmarshal([]byte(s), &current); err != nil {
			t.result = map[string]interface{}{"error": "failed to unmarshal current schema: " + err.Error()}
			return
		}
	default:
		return
	}
	var newSch map[string]interface{}
	if err := json.Unmarshal([]byte(schemaStr), &newSch); err != nil {
		t.result = map[string]interface{}{"error": "failed to unmarshal new schema: " + err.Error()}
		return
	}

	merged := deepMerge(current, newSch)
	mergedData, err := json.Marshal(merged)
	if err != nil {
		t.result = map[string]interface{}{"error": "failed to marshal merged schema: " + err.Error()}
		return
	}
	t.schema = string(mergedData)
}

func (t *NeutralIpcTemplate) HasError() bool {
	if t.result == nil {
		return true
	}
	status, ok := t.result["status"].(byte)
	if !ok || status != CtrlStatusOk {
		return true
	}
	res, ok := t.result["result"].(map[string]interface{})
	if ok {
		if hasError, ok := res["has_error"].(bool); ok && hasError {
			return true
		}
	}
	return false
}

func (t *NeutralIpcTemplate) GetStatusCode() interface{} {
	if t.result == nil {
		return nil
	}
	res, ok := t.result["result"].(map[string]interface{})
	if !ok {
		return nil
	}
	return res["status_code"]
}

func (t *NeutralIpcTemplate) GetStatusText() interface{} {
	if t.result == nil {
		return nil
	}
	res, ok := t.result["result"].(map[string]interface{})
	if !ok {
		return nil
	}
	return res["status_text"]
}

func (t *NeutralIpcTemplate) GetStatusParam() interface{} {
	if t.result == nil {
		return nil
	}
	res, ok := t.result["result"].(map[string]interface{})
	if !ok {
		return nil
	}
	return res["status_param"]
}

func (t *NeutralIpcTemplate) GetResult() map[string]interface{} {
	if t.result == nil {
		return nil
	}
	res, ok := t.result["result"].(map[string]interface{})
	if !ok {
		return nil
	}
	return res
}

// deepMerge merges two maps deeply.
func deepMerge(dict1, dict2 map[string]interface{}) map[string]interface{} {
	merged := make(map[string]interface{})
	for k, v := range dict1 {
		merged[k] = v
	}

	for k, v := range dict2 {
		if val1, ok := merged[k]; ok {
			if m1, ok1 := val1.(map[string]interface{}); ok1 {
				if m2, ok2 := v.(map[string]interface{}); ok2 {
					merged[k] = deepMerge(m1, m2)
					continue
				}
			}
		}
		merged[k] = v
	}
	return merged
}
